uroot/ldd_linux.go

```diff
@@ -0,0 +1,180 @@
+		if err != nil {
+			return fmt.Errorf("%v", err)
+		}
+
+		names[l] = &FileInfo{FullName: l, FileInfo: i}
+		if i.Mode().IsRegular() {
+			return nil
+		}
+		// If it's a symlink, the read works; if not, it fails.
+		// we can skip testing the type, since we still have to
+		// handle any error if it's a link.
+		next, err := os.Readlink(l)
+		if err != nil {
+			return err
+		}
+		// It may be a relative link, so we need to
+		// make it abs.
+		if filepath.IsAbs(next) {
+			l = next
+			continue
+		}
+		l = filepath.Join(filepath.Dir(l), next)
+	}
+}
+
+// runinterp runs the interpreter with the --list switch
+// and the file as an argument. For each returned line
+// it looks for => as the second field, indicating a
+// real .so (as opposed to the .vdso or a string like
+// 'not a dynamic executable'.
+func runinterp(interp, file string) ([]string, error) {
+	var names []string
+	o, err := exec.Command(interp, "--list", file).Output()
+	if err != nil {
+		return nil, err
+	}
+	for _, p := range strings.Split(string(o), "\n") {
+		f := strings.Split(p, " ")
+		if len(f) < 3 {
+			continue
+		}
+		if f[1] != "=>" || len(f[2]) == 0 {
+			continue
+		}
+		names = append(names, f[2])
+	}
+	return names, nil
+}
+
+// Ldd returns a list of all library dependencies for a
+// set of files, suitable for feeding into (e.g.) a cpio
+// program. If a file has no dependencies, that is not an
+// error. The only possible error is if a file does not
+// exist, or it says it has an interpreter but we can't read
+// it, or we are not able to run its interpreter.
+// It's not an error for a file to not be an ELF, as
+// this function should be convenient and the list might
+// include non-ELF executables (a.out format, scripts)
+func Ldd(names []string) ([]*FileInfo, error) {
+	var (
+		list    = make(map[string]*FileInfo)
+		interps = make(map[string]*FileInfo)
+		libs    []*FileInfo
+	)
+	for _, n := range names {
+		r, err := os.Open(n)
+		if err != nil {
+			return nil, err
+		}
+		defer r.Close()
+		f, err := elf.NewFile(r)
+		if err != nil {
+			continue
+		}
+		s := f.Section(".interp")
+		if s == nil {
+			continue
+		}
+		// If there is an interpreter, it should be
+		// an error if we can't read it.
+		i, err := s.Data()
+		if err != nil {
+			return nil, err
+		}
+		// Ignore #! interpreters
+		if len(i) > 1 && i[0] == '#' && i[1] == '!' {
+			continue
+		}
+		// annoyingly, s.Data() seems to return the null at the end and,
+		// weirdly, that seems to confuse the kernel. Truncate it.
+		interp := string(i[:len(i)-1])
+		// We could just append the interp but people
+		// expect to see that first.
+		if interps[interp] == nil {
+			err := follow(interp, interps)
+			if err != nil {
+				return nil, err
+			}
+		}
+		// oh boy. Now to run the interp and get more names.
+		n, err := runinterp(interp, n)
+		if err != nil {
+			return nil, err
+		}
+		for i := range n {
+			if err := follow(n[i], list); err != nil {
+				log.Fatalf("ldd: %v", err)
```

Comment:
I went back and forth on this. But in the end I think that for the use to which this will be put, symlinks should point to something, i.e. be readable. The reason I believe this is simple: for the interpreter to have worked, all the symlinks had to be followable. There are lots of reasons they might not be followable, all of them bad (someone rm'ed something after we ran the interpreter) but that argues for a full rerun of whatever program  is using this function, rather than trying to get too smart.